### ISSUE TITLE
CORS issue with XMLHttpRequest AJAX call

### ISSUE THREAD

**Issue by ls-guillaume-lambert** [DevExpress/testcafe#1680]
### Are you requesting a feature or reporting a bug?

Reporting a bug (coming from `hammerhead.js:8`)

### What is the current behavior?

An AJAX call with XMLHttpRequest to another domain causes CORS issues and returns a `222` status with no response.

### What is the expected behavior?

An AJAX call with XMLHttpRequest should return a response even when its from another domain.

### How would you reproduce the current behavior (if this is a bug)?

Create a test that goes to a page which does an XMLHttpRequest on an external domain.

### Specify your

* operating system: OSX Sierra (10.12.6)
* testcafe version: 0.17.0
* node.js version: 6.11.1

**Comment by ls-guillaume-lambert** [comment]
Continuing on this, when using:

```
nock(/localhost:910[01]/).get('/available', {
    reqheaders: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept'
    }
}).reply(500);
```

It simple times out and never returns anything.

**Comment by miherlosev** [comment]
Hi, @ls-guillaume-lambert .

Specify `reply` headers instead of `reqheaders`.
Also need to add the response body.
See https://github.com/node-nock/nock#specifying-reply-headers.

So, you will need to change your code as follows.
```js
nock(/localhost:910[01]/).get('/available').reply(500, 'reply body', {
        'Access-Control-Allow-Origin': '*'
});
```

Also I've created the example of [testcafe with nock integration](https://github.com/miherlosev/testcafe-nock-integration).

**Comment by ls-guillaume-lambert** [comment]
Yup this does work, but when using the `.reply(function() { ... })` API I still get the `222` code:

```js
nock(/localhost:910[01]/).get('/available').reply((uri, requestBody) => {
    return [500, 'Error', {
        'Access-Control-Allow-Origin': '*'
    }];
);
```

**Comment by ls-guillaume-lambert** [comment]
To put this into context, in our e2e test, we need to change the response when a variable is set:

```js
nock(/localhost:910[01]/).get('/available').reply((uri, requestBody) => {
    return [isAppInstalled ? 200, 500, isAppInstalled ? 'Ok' : 'Error', {
        'Access-Control-Allow-Origin': '*'
    }];
);
```

And in our `testcafe` test:

```js
let isAppInstalled = false;

// nock code above here...

test('simple', async t => {
    await t.click('#someSelector');
    isAppInstalled = true;
    await t.expect(Selector('#installed').exists).ok();
});
```

**Comment by miherlosev** [comment]
'222' is an internal http status code. Testcafe performs validation of a cross-domain request by itself (see https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/request-pipeline/xhr/same-origin-policy.js).

If a request does not pass CORS rules, the proxy responds with 222 status code. 
On the client side, we emulate the native browser behavior: all appropriate getters will return '0' http status code (eg. xhr.statusCode).

In your case, '222' status code means that nock's request mock was configured wrongly and it does not pass CORS rules.
By using `.reply((uri, requestBody) => {}`, you cannot setup response headers and status code. You just return [500, 'Error', {...}] as a response body.
Try to set the debugger on this line https://github.com/node-nock/nock/blob/master/lib\request_overrider.js#L373 and run your tests.

You can also use a working example from the https://github.com/DevExpress/testcafe/issues/1680#issuecomment-321550932.

**Comment by ls-guillaume-lambert** [comment]
Thanks for the response, I will look at debugging this. 

If it fails, do you see other ways of achieving we want as mentioned in https://github.com/DevExpress/testcafe/issues/1680#issuecomment-321566156?

**Comment by miherlosev** [comment]
I will try to understand why nock does not work.

We planned to do it as build-in feature in the nearest release.
See issues:
 https://github.com/DevExpress/testcafe/issues/1341
 https://github.com/DevExpress/testcafe/issues/1271

**Comment by ls-guillaume-lambert** [comment]
Found a way to do it by using `.removeInterceptor()` and re-defining it:

```js
test('Mock XHR Request changes', async t => {
    const xhrNotInstalledInterceptor = nock(/localhost:910[01]/).persist().get('/available');
    const xhrNotInstalledScope = xhrNotInstalledInterceptor.reply(500, 'Error', {
        'Access-Control-Allow-Origin': '*',
    });

    // Do some expects here
    // ...

   nock.removeInterceptor(xhrNotInstalledInterceptor);

    const xhrInstalledInterceptor = nock(/localhost:910[01]/).persist().get('/available');
    const xhrInstalledScope = xhrInstalledInterceptor.reply(200, 'Success', {
        'Access-Control-Allow-Origin': '*',
    });

    // Do more expects here
    // ...
});
```
